cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,392 @@
+
+
+## Simultaneous CID management compatibility
+
+It’s important for Operator and Agents to be compatible when both are managing CIDs at the same time, because upgrading KCP depends on it. This is achieved by moving the CID management functionality to Operator to precisely do the work to get the same results as when Agents are managing CIDs. As long as Agents are managing CIDs, the Operator will only be verifying that the CID and CEP states are correct, without making any modifications. The new CID GC requires no changes in Agents.
+
+The only problem is namespace label changes, because both Operator and Agents will be watching namespaces to react on namespace label changes and create new CIDs and assign them to CEPs as soon as possible. Namespace label change will initially generate duplicate CIDs created from Agents (happens anyway) and Operator (additionally). Operator will receive CEP updates made by Agents and will accept them. Operator will later clean up unused CIDs.
+
+
+## CID deduplication
+
+Operator will not create [duplicate CIDs](#bookmark=id.d4gursbs4oxz), but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+After migration to Operator managing CIDs is completed, deduplication logic should be added as an additional measure of safety.
+
+
+## KVStore compatibility
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with `identity-allocation-mode=CRD`, which is the [default](https://github.com/cilium/cilium/blob/730cc49be6f500ba6b567fd782db3e58e0fa302e/install/kubernetes/cilium/values.yaml#L1508).
+
+All features that rely on using kvstore will not work until there is an equivalent kvstore implementation of Operator managing CIDs.
```

Comment:
If you mean graduate it to stable, meaning making it enabled by default, I agree. But if you mean to be an opt-in feature, I disagree. We want to make smaller steps to continue pushing this forward. And rather than requiring making it compatible with all the features, we would make it opt-it, and until we can ensure that it works with all features, we would clearly document if there are cases which aren't compatible. There are many incompatible config sets in Cilium that would stop cilium-agent from starting.

My point here is that this is a gradual change, and in the first implementation we aren't adding kvstore support. But it is something that shouldn't be a big effort. Perhaps we should track it right from the start?

Reply:
:+1: yeah I think we are in alignment on this. I agree this doesn't need to be a base consideration for initial implementation. I just want to make sure that we do bring this up, consider and discuss, and have a general plan for maturity of the feature regarding stuff like this.

More specifically I think the context I don't understand from the current text is: What does it mean that "All features that rely on using kvstore will not work"? If we were to for instance define the three core pillars of Cilium functionality as networking, security and observability, then I think that security and observability would both be inhibited if not outright broken with the proposal in the current state. For that matter, I think that even basic connectivity would be broken if agent allocation is disabled and kvstore is enabled.

As a matter of maturity I agree we have a number of features in this state and it's not a requirement to make everything work with everything. But yes if we're looking to make this the default long term this would need to be addressed before we reach that point. Also worth noting that although we have tolerated many partial features that are a bit incompatible with one another, this is also creating a burden on development and on usability. There's a broader community discussion to be had amongst contributors on how we manage this complexity across the featureset. Ideally over time we will build up the core architecture to better enable feature compatibility so that this is less of a concern (though in this case inherently there would need to be two codepaths - one for CRD and one for kvstore, the best we can do is minimize duplication). Despite bringing these points up this topic in context of this CFP, it's not necessarily something to be solved within scope of the initial implementation here. I just want to make sure we have aligned expectations and awareness of technical challenges facing the project. Then we can plan collectively to improve things over time. :rocket: